sound/soc/soc-component.c

```diff
@@ -297,39 +297,35 @@ EXPORT_SYMBOL_GPL(snd_soc_component_set_jack);
 int snd_soc_component_module_get(struct snd_soc_component *component,
 				 int upon_open)
 {
-	if (component->module)
+	if (component->module++)
 		return 0;
 
 	if (component->driver->module_get_upon_open == !!upon_open &&
 	    !try_module_get(component->dev->driver->owner))
 		return -ENODEV;
 
-	component->module = 1;
-
 	return 0;
 }
 
 void snd_soc_component_module_put(struct snd_soc_component *component,
 				  int upon_open)
 {
-	if (component->module &&
+	if (component->module-- &&
 	    component->driver->module_get_upon_open == !!upon_open)
 		module_put(component->dev->driver->owner);
-
-	component->module = 0;
 }
 
 int snd_soc_component_open(struct snd_soc_component *component,
 			   struct snd_pcm_substream *substream)
 {
 	int ret = 0;
 
-	if (!component->opened &&
+	if (component->opened[substream->stream] &&
```

Comment:
should be 	if (!component->opened[substream->stream]

Reply:
@plbossart Heh, with this fixes, playback fails again...

Reply:
but if I add this negation, then things don't work for me again. So it looks like the open needs to happen the second time this routine is called?

Reply:
right, so the state machine looks very odd here...

Reply:
@kv2019i what happens when there are multiple capture/playback streams opened? I feel like the opened state is a bad idea

Reply:
@ranj063 Ack. The substreams can be multiple playback streams, so my first patch worked by accident.

I think only sensible thing is to revert the whole thing. I uploaded the patch now, and will sent to the list in a minute.